### PR TITLE
test framework: merge pilot component into istio component

### PR DESCRIPTION
**Pull request by stevenctl**
pilot part of https://github.com/istio/istio/issues/25577

Will make the move to multicluster testing nicer in cases where we check discovery. Rather than needing to initialize many pilot instances we can just get the discovery instance in the site where we actually use it. 

- Adds methods to the `istio.Instance` interface for getting a `Discovery` for a given cluster
- Removed a lot of unused pilots being setup 

- [X] istio.Instance methods to get discovery client
- [X] fix usages
- [X] remove pilot package

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[X] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pkg/test/framework/components/istio/discovery.go [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+	kube2 "istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/test"
+	"istio.io/istio/pkg/test/framework/resource"
+	"istio.io/istio/pkg/test/kube"
+	"istio.io/istio/pkg/test/scopes"
+)
+
+// Discovery allows interacting with the discovery server in a cluster to check for configuration correctness.
+type Discovery interface {
+	CallDiscovery(req *envoyservicediscoveryv3.DiscoveryRequest) (*envoyservicediscoveryv3.DiscoveryResponse, error)
```
Comment (nmittler) [inline pkg/test/framework/components/istio/discovery.go:40]: nit: we should probably rename these methods to be more clear eventually. For example, `StartDiscovery` could be interpreted as starting a discovery server.

File: pkg/test/framework/components/istio/istio.go [hunk 2]
```diff
@@ -15,15 +15,20 @@
...
 
 import (
+	"istio.io/istio/pkg/test"
 	"istio.io/istio/pkg/test/framework/components/environment/kube"
 	"istio.io/istio/pkg/test/framework/resource"
 	"istio.io/istio/pkg/test/scopes"
 )
 
-// Instance represents a deployed Istio instance
+// Discovery represents a deployed Istio instance
```
Comment (nmittler) [inline pkg/test/framework/components/istio/istio.go:24]: Rename fail?

File: pkg/test/framework/components/istio/operator.go [hunk 3]
```diff
@@ -69,7 +71,7 @@ var _ io.Closer = &operatorComponent{}
 var _ Instance = &operatorComponent{}
 var _ resource.Dumper = &operatorComponent{}
 
-// ID implements resource.Instance
+// ID implements resource.Discovery
```
Comment (nmittler) [inline pkg/test/framework/components/istio/operator.go:74]: Here too

File: pkg/test/framework/resource/environment.go [hunk 4]
```diff
@@ -32,6 +32,10 @@ type Environment interface {
 
 	// Clusters in this Environment. There will always be at least one.
 	Clusters() Clusters
+
+	// GetControlPlaneCluster returns the cluster running the control plane for the given cluster based on the ControlPlaneTopology.
+	// An error is returned if the given cluster isn't present in the topology, or the cluster in the topology isn't in KubeClusters.
+	GetControlPlaneCluster(cluster Cluster) (Cluster, error)
```
Comment (nmittler) [inline pkg/test/framework/resource/environment.go:38]: nit: Could we rename this to `GetPrimaryCluster` throughout?

File: pkg/test/framework/components/istio/discovery.go [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+	if d := i.discovery[cluster.Index()]; d != nil {
+		return d, nil
+	}
+
+	return nil, fmt.Errorf("no service discovery client for %s", cluster.Name())
+}
+
+// DiscoveryOrFail returns the service discovery client for the given cluster or the first configured
+// cluster in the environment if given nil. The test will be failed if there is an error.
+func (i *operatorComponent) DiscoveryOrFail(f test.Failer, cluster resource.Cluster) Discovery {
```
Comment (nmittler) [inline pkg/test/framework/components/istio/discovery.go:70]: Maybe we should just get rid of the cluster on this interface for now? AFAICT none of the tests care (pass in nil). WDYT?

**Comment by stevenctl** [comment]
Abandoned in favor of https://github.com/istio/istio/pull/25655

**Comment by howardjohn** [comment]
I think we still need part of this. In my PR the pilot component is left around even though it does nothing really

**Pull request closed without merging**